datafusion/core/src/physical_optimizer/sort_pushdown.rs

```diff
@@ -0,0 +1,416 @@
+    let parent_required = requirements.required_ordering.as_deref();
+    const ERR_MSG: &str = "Expects parent requirement to contain something";
+    let err = || DataFusionError::Plan(ERR_MSG.to_string());
+    if let Some(sort_exec) = plan.as_any().downcast_ref::<SortExec>() {
+        let mut new_plan = plan.clone();
+        if !ordering_satisfy_requirement(plan.output_ordering(), parent_required, || {
+            plan.equivalence_properties()
+        }) {
+            // If the current plan is a SortExec, modify it to satisfy parent requirements:
+            let parent_required_expr =
+                make_sort_exprs_from_requirements(parent_required.ok_or_else(err)?);
+            new_plan = sort_exec.input.clone();
+            add_sort_above(&mut new_plan, parent_required_expr)?;
+        };
+        let required_ordering = new_plan
+            .output_ordering()
+            .map(make_sort_requirements_from_exprs);
+        // Since new_plan is a SortExec, we can safely get the 0th index.
+        let child = &new_plan.children()[0];
+        if let Some(adjusted) =
+            pushdown_requirement_to_children(child, required_ordering.as_deref())?
+        {
+            // Can push down requirements
+            Ok(Transformed::Yes(SortPushDown {
+                plan: child.clone(),
+                required_ordering,
+                adjusted_request_ordering: adjusted,
+            }))
+        } else {
+            // Can not push down requirements
+            Ok(Transformed::Yes(SortPushDown::init(new_plan)))
+        }
+    } else {
+        // Executors other than SortExec
+        if ordering_satisfy_requirement(plan.output_ordering(), parent_required, || {
+            plan.equivalence_properties()
+        }) {
+            return Ok(Transformed::Yes(SortPushDown {
+                required_ordering: None,
+                ..requirements
+            }));
+        }
+        // Can not satisfy the parent requirements, check whether the requirements can be pushed down:
+        if let Some(adjusted) = pushdown_requirement_to_children(plan, parent_required)? {
+            Ok(Transformed::Yes(SortPushDown {
+                plan: plan.clone(),
+                adjusted_request_ordering: adjusted,
+                ..requirements
+            }))
+        } else {
+            // Can not push down requirements, add new SortExec:
+            let parent_required_expr =
+                make_sort_exprs_from_requirements(parent_required.ok_or_else(err)?);
+            let mut new_plan = plan.clone();
+            add_sort_above(&mut new_plan, parent_required_expr)?;
+            Ok(Transformed::Yes(SortPushDown::init(new_plan)))
+        }
+    }
+}
+
+fn pushdown_requirement_to_children(
+    plan: &Arc<dyn ExecutionPlan>,
+    parent_required: Option<&[PhysicalSortRequirement]>,
+) -> Result<Option<Vec<Option<Vec<PhysicalSortRequirement>>>>> {
+    const ERR_MSG: &str = "Expects parent requirement to contain something";
```

Comment:
This structure is not easy to read and understand. I think we may need some future refactoring on this.

`Result<Option<Vec<Option<Vec<PhysicalSortRequirement>>>>>`

Reply:
This will be somewhat simplified when we use the type aliases we tested above. But I agree that we will probably simplify/refactor this code over time